tests/unit/modules/test_datastore.py

```diff
@@ -0,0 +1,253 @@
+        client = mocker.Mock()
+        client.get.return_value = (
+            http.Response(200, '{"spec": {"my": "data"}}')
+        )
+
+        changed, object = datastore.sync(
+            "present", client, "/list", "/resource", {"spec": {"my": "data"}},
+            False,
+        )
+
+        assert changed is False
+        assert {"spec": {"my": "data"}} == object
+        client.put.assert_not_called()
+
+    def test_present_current_object_does_not_differ_check(self, mocker):
+        client = mocker.Mock()
+        client.get.return_value = (
+            http.Response(200, '{"spec": {"my": "data"}}')
+        )
+
+        changed, object = datastore.sync(
+            "present", client, "/list", "/resource", {"spec": {"my": "data"}},
+            True,
+        )
+
+        assert changed is False
+        assert {"spec": {"my": "data"}} == object
+        client.put.assert_not_called()
+
+    def test_present_no_current_object_empty_backend(self, mocker):
+        client = mocker.Mock()
+        client.get.side_effect = (
+            http.Response(404, ""),
+            http.Response(200, "[]"),
+            http.Response(200, '{"spec": {"new": "data"}}'),
+        )
+        client.put.return_value = http.Response(201, "")
+
+        changed, object = datastore.sync(
+            "present", client, "/list", "/resource", {"spec": {"my": "data"}},
+            False,
+        )
+
+        assert changed is True
+        assert {"spec": {"new": "data"}} == object
+        client.put.assert_called_once_with(
+            "/resource", {"spec": {"my": "data"}},
+        )
+
+    def test_present_no_current_object_empty_backend_check(self, mocker):
+        client = mocker.Mock()
+        client.get.side_effect = (
+            http.Response(404, ""),
+            http.Response(200, "[]"),
+        )
+
+        changed, object = datastore.sync(
+            "present", client, "/list", "/resource", {"spec": {"my": "data"}},
+            True,
+        )
+
+        assert changed is True
+        assert {"spec": {"my": "data"}} == object
+        client.put.assert_not_called()
+
+    @pytest.mark.parametrize("check", [False, True])
+    def test_present_no_current_object_non_empty_backend(self, mocker, check):
+        client = mocker.Mock()
+        client.get.side_effect = (
+            http.Response(404, ""),
+            http.Response(200, "[{}]"),
+        )
+
+        with pytest.raises(errors.Error, match="already active"):
+            datastore.sync(
+                "present", client, "/list", "/resource",
+                {"spec": {"my": "data"}}, check,
+            )
+
+        client.put.assert_not_called()
+
+
+class TestDatastore(ModuleTestCase):
+    def test_minimal_datastore_parameters(self, mocker):
+        sync_mock = mocker.patch.object(datastore, "sync")
+        sync_mock.return_value = True, {}
+        set_module_args(
+            name="test_datastore",
+            dsn="my-dsn",
```

Comment:
According to the current argument spec, the minimal set of parameters includes only `name`. If you omitted `dsn` here, the validation wouldn't prevent you from doing that because currently `dsn` is not required for state = present but for state = enabled.

Reply:
I completely forgot that we have two minimal sets of parameters. I added another test for `state == "absent"`. Great catch here.